### PR TITLE
fix: small lock improvement

### PR DESCRIPTION
**Pull request by vansangpfiev**
- Reduce lock scope
- Use atomic for gen_id

### CODE REVIEW

File: src/llama_server_context.cc [hunk 1]
```diff
@@ -707,9 +708,9 @@ bool LlamaServerContext::UpdateSlots() {
           slot.state = SlotState::kProcessing;
           slot.command = SlotCommand::kNone;
           slot.Release();
-          // SendError(slot,
-          //            "Input prompt is too big compared to KV size. Please "
-          //            "try increasing KV size.");
```
Comment (CameronNg) [inline src/llama_server_context.cc:712]: Why were these lines commented out? 🤔

File: src/llama_server_context.cc [hunk 2]
```diff
@@ -1546,10 +1566,15 @@ int LlamaServerContext::SplitMultipromptTask(TaskServer& multiprompt_task) {
...
-  while (!queue_tasks.empty()) {
+  while (true) {
+    std::unique_lock<std::mutex> l(mutex_tasks);
+    if (queue_tasks.empty()) {
+      l.unlock();
+      break;
+    }
     TaskServer task = queue_tasks.front();
     queue_tasks.erase(queue_tasks.begin());
+    l.unlock();
```
Comment (CameronNg) [inline src/llama_server_context.cc:1577]: Is this `unlock` for multiple model 
or just sharing the `queue_tasks` through many threads 🤔

File: src/llama_server_context.cc [hunk 3]
```diff
@@ -258,12 +257,14 @@ int LlamaServerContext::RequestCompletion(json data, bool infill,
...
-    lock.unlock();  // entering new func scope
     return SplitMultipromptTask(task);
   }
 
   // otherwise, it's a single-prompt task, we actually queue it
-  queue_tasks.push_back(task);
+  {
+    std::lock_guard<std::mutex> lock(mutex_tasks);
+    queue_tasks.push_back(task);
+  }
```
Comment (CameronNg) [inline src/llama_server_context.cc:267]: Nice, very elegant way to lock the queue! 👍

File: src/llama_server_context.cc (hunk 1 above)
Comment (vansangpfiev) [reply src/llama_server_context.cc:712]: We did not support `SendError` for slot before, so somehow I commented out those lines :(

**Review by hiento09 — APPROVED** [review]

**Pull request merged**